### PR TITLE
feat: oci genai

### LINKED ISSUE #1049: [Feature]: Add Oracle Cloud Infrastructure (OCI) Generative AI Service as a backend
### Checklist

- [X] I've searched for similar issues and couldn't find anything matching
- [X] I've discussed this feature request in the [K8sGPT Slack](https://k8sgpt.slack.com) and got positive feedback

### Is this feature request related to a problem?

No

### Problem Description

_No response_

### Solution Description

Add an implementation for  [OCI Generative AI Service](https://www.oracle.com/artificial-intelligence/generative-ai/generative-ai-service/) as a k8sgpt backend

### Benefits

Adding the OCI Generative AI backend will provide more backend options to downstream users.

### Potential Drawbacks

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by anders-swanson**
Closes #1049 

## 📑 Description
Adds Oracle Cloud Infrastructure (OCI) GenAI service as a backend for k8sgpt. 

## ✅ Checks

- [] My pull request adheres to the code style of this project
- [] My code requires changes to the documentation
- [] I have updated the documentation as required
- [] All the tests have passed

## ℹ Additional Information

CompartmentId is added as a new parameter for OCI GenAI.

### CODE REVIEW

**Review by AlexsJones — APPROVED** [review]
Thanks for this contribution! Would you please add an example for this in the README under the <details> tab and/or a docs contribution also?  https://github.com/k8sgpt-ai/docs/issues

**Comment by anders-swanson** [comment]
> Thanks for this contribution! Would you please add an example for this in the README under the
> tab and/or a docs contribution also? https://github.com/k8sgpt-ai/docs/issues

Hi Alex,
I'll create a docs PR for this feature.

**Comment by anders-swanson** [comment]
semantic checks seem to be passing now, had to force push the branch.

**Comment by anders-swanson** [comment]
@AlexsJones FYI https://github.com/k8sgpt-ai/docs/pull/105

**Review by AlexsJones — APPROVED** [review]

**Comment by AlexsJones** [comment]
More of an FYI, but if you want this to work in the operator there need to be some changes

**Comment by AlexsJones** [comment]
@k8sgpt-ai/k8sgpt-maintainers  please review

File: pkg/ai/ocigenai.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+				MaxTokens:   &c.maxTokens,
+				Temperature: &temperatureF64,
+				TopP:        &topPF64,
+			},
+		},
+	}
+}
+
+func extractGeneratedText(llmInferenceResponse generativeaiinference.LlmInferenceResponse) (string, error) {
+	response, ok := llmInferenceResponse.(generativeaiinference.CohereLlmInferenceResponse)
```
Comment (JuHyung-Son) [inline pkg/ai/ocigenai.go:86]: I don't know much about OCI. Is there a reason to use Cohere responses and Cohere Request
?

File: pkg/ai/ocigenai.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	if !ok {
+		return "", errors.New("failed to extract generated text from backed response")
+	}
+	sb := strings.Builder{}
+	for _, text := range response.GeneratedTexts {
```
Comment (JuHyung-Son) [inline pkg/ai/ocigenai.go:91]: is this stream?

**Review by JuHyung-Son — COMMENTED** [review]
I haven't done any actual testing, but the code looks fine. I asked a question I don't know much about OCI.

File: pkg/ai/ocigenai.go (hunk 1 above)
Comment (anders-swanson) [reply pkg/ai/ocigenai.go:86]: Right now OCI supports llama3 and cohere, or byo model in cohere format. I would like to add the llama3 support too here, in a follow up.

File: pkg/ai/ocigenai.go (hunk 2 above)
Comment (anders-swanson) [reply pkg/ai/ocigenai.go:91]: By default, stream isn't used. If desired, we can set the IsStream flag in the inference request, so tokens will be sent back as they become available

**Comment by anders-swanson** [comment]
> I haven't done any actual testing, but the code looks fine. I asked a question I don't know much about OCI.

Let me know if you would like to do any OCI testing, I can assist with this.

**Comment by anders-swanson** [comment]
> More of an FYI, but if you want this to work in the operator there needs to be an addition in the supported backends

Is this something I update in this repo, or in the operator repo? Having the operator support could be good 👍

**Pull request merged**